### PR TITLE
Fixes #29144 - Use systemd socket activation

### PR DESCRIPTION
**Pull request by ekohl**
From https://github.com/puma/puma/blob/master/docs/systemd.md#socket-activation
> systemd and puma also support socket activation, where systemd opens the listening socket(s) in advance and provides them to the puma master process on startup. Among other advantages, this keeps listening sockets open across puma restarts and achieves graceful restarts, including when upgraded puma, and is compatible with both clustered mode and application preload.

When using unix sockets, it should also increase security since only the webserver can connect to it. This is not configured by default since the webserver is not assumed to be present. However, it can be deployed via overrides.

This also needs a packaging change for ExecStart. Currently RPM and Deb packaging change this, but the wrappers break the passing on the socket. That part is currently untested so this is still a draft.

### CODE REVIEW

File: extras/systemd/foreman.socket [hunk 1]
```diff
@@ -0,0 +1,21 @@
+[Unit]
```
Comment (ehelms) [inline extras/systemd/foreman.socket:1]: Does the socket log to the syslog? Should we set a `SyslogIdentifier=foreman-socket` on it?

Comment (ekohl) [reply extras/systemd/foreman.socket:1]: No, a socket doesn't really log anything:
```console
# journalctl -u foreman.socket
-- Logs begin at Fri 2020-03-20 11:35:32 UTC, end at Fri 2020-03-20 16:13:38 UTC. --
Mar 20 11:35:34 pipeline-upgrade-foreman-nightly-centos7.wisse.example.com systemd[1]: Listening on Puma HTTP Server Accept Sockets.
```

Now that I see this, the `Description` can be better by using Foreman instead of Puma.

File: extras/systemd/foreman.service [hunk 2]
```diff
@@ -2,13 +2,14 @@
...
 After=network.target remote-fs.target nss-lookup.target
+Requires=foreman.socket
 
 [Service]
 Type=simple
 User=foreman
 TimeoutSec=300
 WorkingDirectory=/usr/share/foreman
-ExecStart=/usr/bin/rails server --environment $FOREMAN_ENV --port $FOREMAN_PORT --binding $FOREMAN_BIND
+ExecStart=/usr/share/foreman/bin/rails server --environment $FOREMAN_ENV --port $FOREMAN_PORT --binding $FOREMAN_BIND
```
Comment (sthirugn) [inline extras/systemd/foreman.service:12]: Why is this change required?

Comment (ekohl) [reply extras/systemd/foreman.service:12]: It was needed to avoid a Bad File Descriptor error that I couldn't figure out. In packaging we also change it to wrap it in `scl enable tfm` which breaks socket activation as well. This script has `/usr/bin/tfm-ruby` as the shebang in RPM packaging and works.

**Comment by sthirugn** [comment]
I tested this change and it still took about 16 seconds between `Started Foreman` (line 2 below - Mar 20 18:21:12) and the actual puma initialization (last line below - 18:21:28). Wouldn't this change improve the startup time?

```console
Mar 20 18:20:55 my-nightly-koji.apex.example.com systemd[1]: Stopped Foreman.
Mar 20 18:21:12 my-nightly-koji.apex.example.com systemd[1]: Started Foreman.
Mar 20 18:21:15 my-nightly-koji.apex.example.com foreman[27599]: /usr/share/foreman/lib/foreman.rb:8: warning: already initialized constant Foreman::UUID_REGEXP
Mar 20 18:21:15 my-nightly-koji.apex.example.com foreman[27599]: /usr/share/foreman/lib/foreman.rb:8: warning: previous definition of UUID_REGEXP was here
Mar 20 18:21:16 my-nightly-koji.apex.example.com foreman[27599]: => Booting Puma
Mar 20 18:21:16 my-nightly-koji.apex.example.com foreman[27599]: => Rails 5.2.1 application starting in production
Mar 20 18:21:16 my-nightly-koji.apex.example.com foreman[27599]: => Run `rails server -h` for more startup options
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] Puma starting in cluster mode...
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Version 3.11.4 (ruby 2.5.5-p157), codename: Love Song
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Min threads: 0, max threads: 16
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Environment: production
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Process workers: 2
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Phased restart available
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] * Listening on tcp://127.0.0.1:3000
Mar 20 18:21:27 my-nightly-koji.apex.example.com foreman[27599]: [27614] Use Ctrl-C to stop
Mar 20 18:21:28 my-nightly-koji.apex.example.com foreman[27599]: [27614] - Worker 0 (pid: 27656) booted, phase: 0
Mar 20 18:21:28 my-nightly-koji.apex.example.com foreman[27599]: [27614] - Worker 1 (pid: 27664) booted, phase: 0
```

**Comment by ekohl** [comment]
Try performing a a request to Apache during those 14 seconds. Before this change it would be returns a HTTP 503. With this change it blocks until Foreman has started. That's the entire goal of this change.

**Comment by sthirugn** [comment]
> Try performing a a request to Apache during those 14 seconds. Before this change it would be returns a HTTP 503. With this change it blocks until Foreman has started. That's the entire goal of this change.

Thanks, I couldn't test the entire setup due to the complexity in setting up the test server. But the change looks sane to me.

**Review by ehelms — APPROVED** [review]
